Fix #3022 - add melter toggle

dnil opened the pull request:
This PR adds a functionality or fixes a bug.

![Screenshot 2021-12-03 at 16 26 26](https://user-images.githubusercontent.com/758570/144628244-62e5ba9e-f8fb-485d-aa51-13a982c6f285.png)

![Screenshot 2021-12-03 at 17 14 09](https://user-images.githubusercontent.com/758570/144635455-cb62ef28-2a1e-4870-8a00-33245a1f5060.png)

![Screenshot 2021-12-03 at 17 31 18](https://user-images.githubusercontent.com/758570/144638242-c70e6c05-087e-447c-b106-e4c79c827742.png)

![Screenshot 2021-12-03 at 17 39 10](https://user-images.githubusercontent.com/758570/144639520-a2b6db36-9b01-4a1b-a94f-01ff61525be7.png)

**How to test**:
1. toggle rerun monitoring on and off for cases
2. note that this should be logged as events for the case
3. when rerun monitoring is enabled for a case, it should be exported ok using `scout export cases --rerun-monitor`

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [ ] code approved by
- [ ] tests executed by

dnil commented:
This is working, but I will add a couple of tests to get coverage back up again.

henningonsbring commented on scout/server/blueprints/cases/templates/cases/collapsible_actionbar.html:
```diff
@@ -275,6 +275,11 @@
   {% endif %}
+  <p>
+    <form action="{{ url_for('cases.monitor', institute_id=institute._id, case_name=case.display_name) }}" method="POST">
+      <span><input type="checkbox" id="rerun_monitoring" onChange="this.form.submit()" name="rerun_monitoring" value="monitor" {% if case.rerun_monitoring %}checked{% endif %}></span><span class="text-white" data-toggle="popover" data-content="Check box to enable rerun monitoring with melter."> Request rerun monitoring</span>
```
```suggestion
      <span><input type="checkbox" id="rerun_monitoring" onChange="this.form.submit()" name="rerun_monitoring" value="monitor" {% if case.rerun_monitoring %}checked{% endif %}></span><span class="text-white" data-toggle="popover" data-content="A new analysis will start when relevant parts of the workflow is updated"> Activate monitoring</span>
```

henningonsbring commented on scout/commands/export/case.py:
```diff
@@ -21,6 +21,7 @@
 @click.option("-f", "--finished", is_flag=True, help="archived or solved")
 @click.option("--causatives", is_flag=True, help="Has causative variants")
 @click.option("--research-requested", is_flag=True, help="If research is requested")
+@click.option("--rerun-monitor", is_flag=True, help="If case rerun monitoring is requested")
```
"If case rerun monitoring is requested", from just reading this text I do not really understand what it does, can it be more clear? Now it sounds like you ask if a specific case is monitored, as I understand it you will get all monitored cases returned together with some metadata.

dnil replied:
Yeah, I just followed the style we have for that endpoint, but I agree; it's not easy to parse without examples. I've added a few words; hopefully it's a little better.

dnil commented:
Ok, fixed conflict. I have now added a config option to show the melter interaction in the web interface, so that we could merge the code without having to think too much about how the users will receive it when the melter backend is not yet up in production.

There is one thing that will require a little interaction testing with cg: the switch to our own `json_converter` instead of the `bson_handler`. I think it is the way to do it. Using the `bson_handler` is actually bugged on the current `main`. Export case with PNGs stored, like for the demo, with the `--json` flag crashes. It might still be that some cg component is surprised - e.g. if it expects the broken 'datetime.datetime()`-strings in the output. Probably not, but better safe than sorry.

northwestwitch commented on .github/workflows/vulture.yml:
```diff
@@ -19,5 +19,5 @@ jobs:
         id: vulture
         with:
-          vulture-args: --min-confidence 80 --ignore-names cls,args ${{steps.files.outputs.all}}
+          vulture-args: --min-confidence 80 --ignore-names cls,args,kwargs ${{steps.files.outputs.all}}
```
👍🏻

northwestwitch commented on CHANGELOG.md:
```diff
@@ -80,6 +80,8 @@ About changelog [here](https://keepachangelog.com/en/1.0.0/)
 - Add utils for dictionary type
 - Filter institute cases using multiple HPO terms
 - Vulture GitHub action to identify and remove unused variables and imports
+- Add a toggle for melter rerun monitoring of cases
```
It has been a while, so these 2 should be moved up in the changelog

dnil replied:
Indeed! I thought I did that in a merge, but time flies.

northwestwitch commented on scout/adapter/mongo/case_events.py:
```diff
@@ -413,11 +413,73 @@ def request_rerun(self, institute, case, user, link):
+            subject=case["display_name"],
+        )
+
+        updated_case = self.case_collection.find_one_and_update(
```
This works. But one thing that I noticed while working on a recent PR is that in general all case updates go through the [update_case](https://github.com/Clinical-Genomics/scout/blob/c1630a12b53509b779abf7a06fb8666b36842674/scout/adapter/mongo/case.py#L755).

I think the code would be easier to maintain if you update the case in the same way

dnil replied:
Hm, this is still in the case adapter.. But, sure, the fewer places we touch the db the better.

northwestwitch commented on scout/adapter/mongo/case_events.py:
```diff
@@ -413,11 +413,73 @@ def request_rerun(self, institute, case, user, link):
+            subject=case["display_name"],
+        )
+
+        updated_case = self.case_collection.find_one_and_update(
```
Same here, invoke `self.update_case` instead?

northwestwitch commented on scout/commands/export/case.py:
```diff
@@ -7,6 +7,7 @@
 from scout.server.extensions import store
+from scout.server.utils import jsonconverter
 
 from .export_handler import bson_handler
```
I think Vulture is right here, it's not used any more

dnil replied:
Yeah, that likely was the last place.

northwestwitch commented on scout/commands/export/case.py:
```diff
@@ -21,6 +22,11 @@
 @click.option("-f", "--finished", is_flag=True, help="archived or solved")
 @click.option("--causatives", is_flag=True, help="Has causative variants")
 @click.option("--research-requested", is_flag=True, help="If research is requested")
+@click.option(
```
Useful! 👍🏻

northwestwitch commented on scout/server/blueprints/cases/templates/cases/collapsible_actionbar.html:
```diff
@@ -276,16 +276,33 @@
+{% macro melter_interact(institute, case) %}
+  <div class="ml-3 mt-1">
+    <form action="{{ url_for('cases.monitor', institute_id=institute._id, case_name=case.display_name) }}" method="POST">
+      <span><input type="checkbox" id="rerun_monitoring" onChange="this.form.submit()" name="rerun_monitoring" value="monitor" {% if case.rerun_monitoring %}checked{% endif %}></span><span class="text-white" data-toggle="tooltip" title="A new analysis will start when relevant parts of the workflow is updated"> Activate monitoring</span>
```
```suggestion
      <span><input type="checkbox" id="rerun_monitoring" onChange="this.form.submit()" name="rerun_monitoring" value="monitor" {% if case.rerun_monitoring %}checked{% endif %}></span><span class="text-white" data-toggle="tooltip" title="A new analysis will start when relevant parts of the workflow are updated"> Activate monitoring</span>
```

northwestwitch commented on scout/commands/export/case.py:
```diff
@@ -70,7 +80,7 @@ def cases(
 
     if json:
-        click.echo(json_lib.dumps(models, default=bson_handler))
+        click.echo(json_lib.dumps(models, default=jsonconverter))
```
👍🏻

northwestwitch commented on scout/constants/case_tags.py:
```diff
@@ -88,6 +88,8 @@
     "share": "shared case with",
     "unshare": "revoked access for",
     "rerun": "requested rerun of",
+    "monitor": "requested rerun monitoring for",
```
```suggestion
    "rerun_monitor": "requested rerun monitoring for",
```

northwestwitch commented on scout/constants/case_tags.py:
```diff
@@ -88,6 +88,8 @@
     "unshare": "revoked access for",
     "rerun": "requested rerun of",
+    "monitor": "requested rerun monitoring for",
+    "unmonitor": "disabled rerun monitoring for",
```
```suggestion
    "rerun_unmonitor": "disabled rerun monitoring for",
```

northwestwitch commented on scout/adapter/mongo/case_events.py:
```diff
@@ -413,11 +413,73 @@ def request_rerun(self, institute, case, user, link):
+            user=user,
+            link=link,
+            category="case",
+            verb="monitor",
```
```suggestion
            verb="rerun_monitor",
```

northwestwitch commented on scout/adapter/mongo/case_events.py:
```diff
@@ -413,11 +413,73 @@ def request_rerun(self, institute, case, user, link):
+            user=user,
+            link=link,
+            category="case",
+            verb="unmonitor",
```
```suggestion
            verb="rerun_unmonitor",
```

northwestwitch commented on scout/server/blueprints/cases/templates/cases/collapsible_actionbar.html:
```diff
@@ -276,16 +276,33 @@
 </a>
 <!-- Submenu content -->
 {% if case.rerun_requested %}
+  {% if config.SHOW_MELTER %}
```
SHOW_MELTER is a bit tricky as a name for a config param. It's not self-explanatory and it has a meaning just for CG Solna. Rename it to something like `RERUN_MONITORING` ?

northwestwitch commented on scout/server/blueprints/cases/templates/cases/collapsible_actionbar.html:
```diff
@@ -276,16 +276,33 @@
 {% if case.rerun_requested %}
+  {% if config.SHOW_MELTER %}
+    <div id='submenu3' class="collapse">
+      {{ melter_interact(institute, case) }}
```
Also here, perhaps rename to something that can be understood and used also by other centers?

dnil replied:
He, sure, it's just a macro name, so not very exposed. But it doesn't matter, so why not!

northwestwitch commented on scout/server/blueprints/cases/templates/cases/collapsible_actionbar.html:
```diff
@@ -276,16 +276,33 @@
 {% else %}
   <div id='submenu3' class="collapse">
+    {% if config.SHOW_MELTER %}
+      {{ melter_interact(institute, case) }}
```
Here as well

northwestwitch commented on scout/server/blueprints/cases/views.py:
```diff
@@ -828,6 +828,21 @@ def rerun(institute_id, case_name):
 
 
+@cases_bp.route("/<institute_id>/<case_name>/monitor", methods=["POST"])
+def monitor(institute_id, case_name):
```
```suggestion
def rerun_monitor(institute_id, case_name):
```

northwestwitch commented on scout/server/config.py:
```diff
@@ -97,6 +97,7 @@
 SHOW_CAUSATIVES = True
 SHOW_OBSERVED_VARIANT_ARCHIVE = True
 HIDE_ALAMUT_LINK = True
+SHOW_MELTER = True
```
RERUN_MONITORING ?

northwestwitch commented on tests/server/blueprints/cases/test_cases_views.py:
```diff
@@ -91,6 +89,48 @@ def test_reanalysis(app, institute_obj, case_obj, mocker, mock_redirect):
         assert resp.status_code == 302
 
 
+def test_monitor(app, institute_obj, mocker, mock_redirect):
```
```suggestion
def test_rerun_monitor(app, institute_obj, mocker, mock_redirect):
```

northwestwitch commented on tests/server/blueprints/cases/test_cases_views.py:
```diff
@@ -91,6 +89,48 @@ def test_reanalysis(app, institute_obj, case_obj, mocker, mock_redirect):
+
+        # AND an unmonitor event should be created
+        rerun_event = store.event_collection.find_one()
+        assert rerun_event.get("verb") == "monitor"
```
```suggestion
        assert rerun_event.get("verb") == "rerun_monitor"
```

northwestwitch reviewed (commented):
Nice! I have a few suggestions, but nothing so complicated!

What happens when the case is rerun? Then the rerun monitor should be set to False for the case, right?

dnil commented:
> Nice! I have a few suggestions, but nothing so complicated!

Thanks a bunch! I'll check.

> What happens when the case is rerun? Then the rerun monitor should be set to False for the case, right?

No, that doesn't really affect it yet. Maybe the case was solved by the first rerun, maybe it wasn't. Possibly it should be auto reset if the case is marked solved, but then there is partially solved, so I think we can let the user decide for starters.

At some point we will need to present results from specifically the melter reruns. It will likely entail a little bit of tagging of variants of interest to avoid complete re-evaluation of the case. That will be a later PR...

northwestwitch commented on scout/server/config.py:
```diff
@@ -97,6 +97,8 @@
 SHOW_OBSERVED_VARIANT_ARCHIVE = True
 HIDE_ALAMUT_LINK = True
+# Display case rerun monitoring toggle
+RERUN_MONITOR = True
```
```suggestion
# RERUN_MONITOR = True
```

northwestwitch replied:
Just an idea: since other centers are not likely going to have a use for the functionality just now, we could switch if off in the demo settings?

dnil replied:
Hm, this is the demo config. I would suggest keeping it enabled as possible, when other components are not really invoked. I mean, a test user could run the cli export command for monitored cases. But it doesn't matter very much either way.

dnil replied:
Ah, but that reminds me - what happens if the key does not exist! Will fix!

northwestwitch replied:
Ok, see, my comment has been useful anyway! :)

dnil replied:
> Ok, see, my comment has been useful anyway! :)

Essentially always! 😸

northwestwitch reviewed (approved):
Tested again after the changes and works perfectly! 💯
